tools/compare.go

```diff
@@ -15,125 +15,291 @@
 
-	go func() {
-		defer wg.Done()
-		_ = collectBlocks(storeB, blocksB)
-	}()
-	wg.Wait()
+			for _, expectedBlockHash := range expectedBlockHashes {
+				blocksCountedInChunk++
+				expectedBlock := expectedBlocks[expectedBlockHash]
+				receivedBlock, existsInReceived := receivedBlocks[expectedBlockHash]
 
-	for blockID, blockA := range blocksA {
-		blockB, exists := blocksB[blockID]
-		if !exists {
-			continue
-		}
+				// Reset chunk, print if good
+				if blocksCountedInChunk >= 100000 || uint64(blocksCountedInChunk) >= (blockRangeSize-uint64(rangeNum*100000)-1) || uint64(blocksCountedInChunk) == blockRangeSize-1 {
+					if chunkIsGood {
+						fmt.Printf("✓ Bundle %d - %d has no differences\n", (rangeNum)*100000, (rangeNum+1)*100000)
+					}
+					chunkIsGood = true
+					rangeNum++
+					blocksCountedInChunk = -1
+				}
+
+				// false && first error in chunk
+				if !existsInReceived && chunkIsGood {
+					chunkIsGood = false
+					fmt.Printf("✖ Bundle %d - %d is different\n", (rangeNum)*100000, (rangeNum+1)*100000)
+					bundleHasDiff = true
+				}
+
+				// Check if --diff is enabled
+				if displayDiff {
+					if !existsInReceived {
+						fmt.Printf("- Block (%s) is present in %s but missing in %s\n", expectedBlock.AsRef(), args[0], args[1])
+						bundleHasDiff = true
+					}
+					if !proto.Equal(expectedBlock, receivedBlock) {
+						bundleHasDiff = true
+
+						if chunkIsGood {
+							chunkIsGood = false
+							fmt.Printf("✖ Bundle %d - %d is different\n", (rangeNum)*100000, (rangeNum+1)*100000)
+						}
+						fmt.Printf("- Block (%s) is different\n", expectedBlock.AsRef())
+
+						expectedBlockJSON, err := rpc.MarshalJSONRPCIndent(expectedBlock, "", " ")
+						if err != nil {
+							return fmt.Errorf("marshaling block: %w", err)
+						}
+						receivedBlockJSON, err := rpc.MarshalJSONRPCIndent(receivedBlock, "", " ")
+						if err != nil {
+							return fmt.Errorf("marshaling block: %w", err)
+						}
+
+						diff, err := unifiedDiff(expectedBlockJSON, receivedBlockJSON)
+						if err != nil {
+							return fmt.Errorf("getting diff: %w", err)
+						}
+						fmt.Printf("difference: \n%s\n", diff)
+					}
+				} else {
 
-		blockAJSON, _ := json.Marshal(blockA)
-		blockBJSON, _ := json.Marshal(blockB)
+					// And doesn't exist
+					if !existsInReceived {
+						fmt.Printf("- Block #%d (%s) is different\n", expectedBlock.Number, expectedBlock.Hash)
+						bundleHasDiff = true
 
-		diff := cmp.Diff(blockAJSON, blockBJSON)
-		if diff != "" {
-			fmt.Printf("❌ difference found on block id %s\n", blockID)
-			fmt.Println(diff)
-			return nil
+						// Exists but has diff
+					} else if !proto.Equal(expectedBlock, receivedBlock) {
+						bundleHasDiff = true
+
+						if chunkIsGood {
+							chunkIsGood = false
+							fmt.Printf("✖ Bundle %d - %d is different\n", (rangeNum)*100000, (rangeNum+1)*100000)
+						}
+						fmt.Printf("- Block (%s) is different\n", expectedBlock.AsRef())
+					}
+				}
```

Comment:
This part to me feels like it requires refactoring, really hard to "read". One thing that I note is that there seems to be some "duplicated" logic making the flow harder to reason about, for example `proto.Equal` is called in twice depending on some `--diff`, that seems weird at first sight.

I feel the initial `if displayDiff {` is the problem. A lot of the same logic applies be it that we display the diff or not. The diff only add the `unifiedDiff` in addition of the "default" display, so it feels weird that the `if` is so high in the diffing logic.

Please **do** this refactoring of making this piece more streamlined.